### ISSUE TITLE
[Feature Request] Allow configurable PDBs

### ISSUE THREAD

**Issue by fullykubed** [cloudnative-pg/cloudnative-pg#2570]
**Use Case:** We want to allow eviction of the primary DB instance in order to perform upgrades on the underlying infrastructure automatically, for example updating a kubernetes node. While that will cause degraded performance during the failover, that is aligned with our SLOs for our uses of a majority of the clusters.

**Problem:** The current operator sets up a PDB that never allows the primary to be evicted. No configuration parameters are exposed to override this default behavior.

Edit:

I have expanded on additional use cases and reasons why this is critically important in a follow-up [comment](https://github.com/cloudnative-pg/cloudnative-pg/issues/2570#issuecomment-1858650998).

**Comment by leonardoce** [comment]
Hello.

The operator is watching over nodes. 
When the node where the primary Pod is running is being cordoned, it will automatically switchover to the most advanced replica running on a node which is not cordoned.

This will allow you to evict the Pods that are running on the node you drained, even if the PDB is still there.
Is this enough for you?

**Comment by jclangst** [comment]
Thanks for the response. That is helpful info and confirmed that cordoning works for failovers to a running replica. However, there are a handful of scenarios that we encounter that don't work.

- EKS node upgrades ([docs](https://docs.aws.amazon.com/eks/latest/userguide/managed-node-update-behavior.html#managed-node-update-upgrade)): The cordoning happens _after_ the eviction and causes a `PodEvictionFailure` due to not being able to evict the pg instances. We would like to avoid enabling the `force` parameter globally here as we have other cases where we would want PDBs for other resources to stop a node upgrade.

- Clusters that do not have any replicas still get this PDB. The cordoning strategy does not work in this case. We run some ephemeral  DB clusters without replicas or backups as they are meant to just be temporary.  In this case, the clusters do not need to be highly available and we don't want PDBs at all.

**Comment by jclangst** [comment]
It would be nice to just have a simple boolean parameter on `Cluster` that disables the PDB management behavior in the operator for that particular cluster.

**Comment by leonardoce** [comment]
I think I got it now.
You can set the cluster in maintenance mode, and that will remove the PDB while the maintenance mode is set.
More info:

https://cloudnative-pg.io/documentation/1.20/kubernetes_upgrade/#kubernetes-upgrade

Would this help?

**Comment by jclangst** [comment]
Unfortunately not. We don't want to disable the self-healing features. We want to be able to evict the primary _and_ have the controller self-heal as it is able to. 

For our ephemeral clusters, we want to avoid adding another control loop that has to toggle flags on the Cluster resource just to use the eviction API in our automations.

For our use case, we specifically _want_ to allow disruptions of the primary. The PDB prevents that. That's a great default, but we specifically want to opt-out on a per-cluster basis.

**Comment by vladimirfx** [comment]
Similar situation here. We have many DBs without replicas (only backups) that is ok to shutdown on cluster upgrades. Ability to just disable PDB on cluster basis help us too.

**Comment by tomaszkrzyzanowski** [comment]
Same here, we don't have control over the clusters where the CNPG is running, and its PDB is breaking the K8s cluster upgrades. It's not possible to switch the clusters to maintenance mode – the support is not informing us about K8s upgrades, and we don't have control over infra.

PDB toggle would be much appreciated :)

**Comment by AlfredTM** [comment]
Same here, we are using AKS with Autoupgrade and the PDB stops the upgrade. Need to evict the pod manually for the node to restart.

**Comment by jclangst** [comment]
@leonardoce Does this use case here make sense to add this capability to disable the PDB as a non-default option? We would like to use this project, but this is a blocker.

**Comment by jclangst** [comment]
For anyone curious for a workaround, we settled on running this in a periodic cronjob in the interim. Unfortunately, this is necessary operator will restore the PDB on every cluster update or when the operator pods restart. Additionally, the operator does not respect field managers.

```bash
#!/bin/bash

kubectl get pdb --all-namespaces -o json | jq -r \
  '.items[] | 
   select(.metadata.annotations["cnpg.io/operatorVersion"]? != null) |
   "\(.metadata.namespace) \(.metadata.name)"' | \
while read -r namespace name; do
  kubectl delete pdb "$name" -n "$namespace"
done
```

**Comment by jclangst** [comment]
The saga continues. Apparently, the script above does not work in later versions as the operator is in a realtime watch mode for changes or deletions of the pdb and will restore the pdb immediately after you try to remove them. The only way to buy yourself a bit of breathing room is to keep the pdb, remove the ownerReferences, and the change the `minAvailable` to `0`.

```bash
#!/usr/bin/env bash

# This script will patch all pdbs managed by the cnpg operator to allow primary disruption

set -eo pipefail


kubectl get pdb --all-namespaces -o json | jq -r \
  '.items[] |
   select(.metadata.annotations["cnpg.io/operatorVersion"]? != null) |
   "\(.metadata.namespace) \(.metadata.name)"' | \
while read -r namespace name; do
  # Delete owner references
  kubectl patch pdb "$name" -n "$namespace" --type='json' -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
  # Set minAvailable to 0
  kubectl patch pdb "$name" -n "$namespace" --type='json' -p='[{"op": "replace", "path": "/spec/minAvailable", "value":0}]'
done
```
Granted it will still restore itself occasionally like when the operator restarts, so you have to run this in a cronjob.

**Comment by Christian-Probst-DATEV** [comment]
We need that feature too! Please.

**Comment by jbattiato** [comment]
Hi @jclangst

Could you please elaborate more on how the self healing in a single node cluster would help you while you are evicting the only primary pod?

What are you trying to achieve?

This is how CNPG allows you to upgrade K8S node in case of a single node cluster:
```
nodeMaintenanceWindow:
  inProgress: true
  reusePVC: true
```
as explained by Leonardo a few posts above:
https://cloudnative-pg.io/documentation/1.21/kubernetes_upgrade/

These settings will automatically remove the `PodDisruptionBudget` blocker to allow pod eviction when node drain is requested.

Basically, when you run a single node cluster you accept the downtime while upgrading the Kubernetes node.
The primary pod will get back up and running once the node is ready to schedule workloads again. The primary then reattaches itself to the PVC and starts Postgres.

Would that be enough for you?

**Comment by Christian-Probst-DATEV** [comment]
We have a team doing the maintenance for k8s and our Team is running the database operator. 3 Nodes, 1 Primary, 2 Standbys. Our Team doesn't know when maintenance is taking place and is not involved with it. We need a way to remove, better avoid, all PDB's by declaration.

**Comment by jbattiato** [comment]
> We have a team doing the maintenance for k8s and our Team is running the database operator. 3 Nodes, 1 Primary, 2 Standbys. Our Team doesn't know when maintenance is taking place and is not involved with it. We need a way to remove, better avoid, all PDB's by declaration.

Hi @Christian-Probst-DATEV 

You don't need to remove PDB when a cluster has at least one replica. In this case the PDB does not prevent the drain.
Indeed, the Operator will ensure the primary is always running on a schedulable node. That means, when the node drain kicks in against the node where the primary is running, the Operator will promote the most aligned replica before allowing the eviction of the primary pod.

I just did this test in a `kind` cluster with three (virtual) K8S nodes. I created a CNPG cluster with 2 nodes, one primary and one standby. I have not set the `nodeMaintenanceWindow`, so the default was set to `false`. I did execute `kubectl cordon` against the node where the primary was running. The result was the Operator performed the switchover to promote the standby as the new primary, on the K8S node that was still schedulable.
CNPG version: compiled from main, using the `hack/setup-cluster.sh` script.

So, my questions now are:

* did you face any other blockers? Or
* the Operator did not behave as expected in your specific case?

Please, elaborate more on the issue so we can help you resolve any blockers.

**Comment by jclangst** [comment]
Thanks for the follow-up, @jbattiato!

To summarize what I have said above: we want to be able to turn off the PDB for the primary instance as an optional configuration flag. We want to allow the primary to be disrupted at any arbitrary time utilizing nothing more than the standard kubernetes eviction API. 

Why? All common controllers that perform autoscaling or node maintenance operations are disrupted if a PDB prevents a pod eviction. I can name a few that we utilize: EKS, vertical pod autoscaler, cluster autoscaler, karpenter, descheduler. Others have pointed to some other controllers that their teams utilize as well.

If you would like to drill down into one one of those, I am happy to provide more detail. I have already provided a cursory explanation of the EKS blocker above.

I understand that both you and @leonardoce have suggested the node maintenance mode as a potential solution. That does _not_ work for the following reasons:

- Per the docs you linked, maintenance mode disables self healing. By "self healing," I have assumed this refers to the operator relaunching the primary pod if it is disrupted.  We want _both_ the ability to disrupt the primary _and_ the ability for the operator to recover from the disruption.  As a result, a permanent maintenance mode is a non-starter.

  > "...you are advised to temporarily put the cluster in maintenance mode through the nodeMaintenanceWindow option   to avoid standard self-healing procedures to kick in, while, for example, enlarging the partition on the physical node or updating the node itself... Limit the duration of the maintenance window to the shortest amount of time possible. In this phase, some of the expected behaviors of Kubernetes are either disabled or running with some limitations, including self-healing, rolling updates, and Pod disruption budget." 

- Toggling on and off node maintenance mode during temporary maintenance windows is also a non-starter.  We are looking to accommodate processes that run continuously, not just at a point-in-time like a manual k8s node upgrade. As just one example, the cluster autoscaler will ignore nodes with non-disruptible pods during its autoscaling calculations.  In our use cases, we want to empower the controllers that I listed above to disrupt any CNPG cluster at any arbitrary time. We care MUCH more about this than temporary service disruptions to a CNPG cluster.

  - To provide more context, we are considering running 100s of CNPG clusters in our testing, integration, and individual developer environments. This is a non-starter for cost reasons when CNPG effectively blocks automatic node maintenance and autoscaling due to this PDB.

- In a CNPG cluster _with_ replicas, maintenance mode doesn't remove the PDB at all.

I hope this has provided the additional necessary context for the "Why" of this feature request.

**Comment by Christian-Probst-DATEV** [comment]
> > We have a team doing the maintenance for k8s and our Team is running the database operator. 3 Nodes, 1 Primary, 2 Standbys. Our Team doesn't know when maintenance is taking place and is not involved with it. We need a way to remove, better avoid, all PDB's by declaration.
> 
> Hi @Christian-Probst-DATEV
> 
> You don't need to remove PDB when a cluster has at least one replica. In this case the PDB does not prevent the drain. Indeed, the Operator will ensure the primary is always running on a schedulable node. That means, when the node drain kicks in against the node where the primary is running, the Operator will promote the most aligned replica before allowing the eviction of the primary pod.
> 
> I just did this test in a `kind` cluster with three (virtual) K8S nodes. I created a CNPG cluster with 2 nodes, one primary and one standby. I have not set the `nodeMaintenanceWindow`, so the default was set to `false`. I did execute `kubectl cordon` against the node where the primary was running. The result was the Operator performed the switchover to promote the standby as the new primary, on the K8S node that was still schedulable. CNPG version: compiled from main, using the `hack/setup-cluster.sh` script.
> 
> So, my questions now are:
> 
> * did you face any other blockers? Or
> * the Operator did not behave as expected in your specific case?
> 
> Please, elaborate more on the issue so we can help you resolve any blockers.

@jbattiato thanks for reply, the k8s team has tested your advice with success. Works as designed.

**Comment by fcanovai** [comment]
This seems mostly a documentation and naming issue to me, so let me first summarise the operator behaviour, according to the Cluster configuration and storage.

### Single instance cluster

The `$CLUSTERNAME-primary` PDB exists by default to make sure the primary is not impacted by mistake.

On a single instance cluster, permanently keeping `nodeMaintenanceWindow.inProgress: true` and `nodeMaintenanceWindow.reusePVC: true` will not have any other effect except disabling the `$CLUSTERNAME-primary` PDB. They can be set at cluster creation time.

### Multiple instances cluster

On a multiple instance cluster the behaviour may vary, depending on the possibility to reattach the storage on other nodes and the role of the pods included running in the node to be drained.

1. Storage can be reattached
    
    No need to set the anything in the `nodeMaintenanceWindow`. A pod will be restarted on a different node with the same PVC.
1. Storage can not be reattached
    
    If the storage is local to the node, there are multiple possibilities.
    1. `reusePVC: false && inProgress: true`
    
        Pods on the node will be deleted together with their PVCs, new pods will be created backing up the primary. Not suggested for large databases, since the backup could take longer than the actual maintenance.
    1. `reusePVC: true && inProgress: true`
    
        The `$CLUSTERNAME` pdb gets deleted. This PDB prevents more than a single standby to be down, so deleting it will allow you to drain a node with multiple standbys.
    The pods will be deleted by the drain, and they will stay `Pending` until the node is available again. Only at this point, when a pod is unschedulable, some of the self-healing features won't be available. For example, if another pod of the cluster is deleted, it won't be recreated, but if that pod is the primary and another pod exists, another one will be promoted.
    When you uncordon the node and the pods are schedulable again, everything will be back to normal. So, like in the single instance scenario, this configuration can exist permanently. The cluster will miss the `$CLUSTERNAME` pdb (not the `$CLUSTERNAME-primary`, but that can't block a drain in this case).
    1. `nodeMaintenanceWindow` not set or `inProgress: false`
        
        Not suggested. The `$CLUSTERNAME` pdb is not removed, so if there are multiple standbys on the same node the drain can get stuck.

---

The documentation is more threatening than it should be, and deserves a change (tied to https://github.com/cloudnative-pg/cloudnative-pg/issues/3557).

@Christian-Probst-DATEV , @jclangst : are your use cases covered by the above or is there something else?

**Comment by gbartolini** [comment]
We need to do something here indeed. The whole maintenance window stuff was created a long time ago with some specific and restricted use case. The problem stated in this issue is real and we need to find a reasonable enhancement, even though if it means changing or extending the API (being careful not to break existing behaviours).

**Comment by jclangst** [comment]
Unfortunately, the suggestions provided by [fcanovai](https://github.com/fcanovai) do not remove all of the underlying PDBs. In particular, still cannot seem to get rid of the PDB for the primary instance in a multi-node setup.

I cannot speak to the history or holistic impact of turning on the maintenance window, but I can say that this does not seem to semantically map to what we are trying to accomplish. The only thing desired is an option to disable _all_ PDBs _without_ having any other impact to the cluster behavior.

Even if there was some way to accomplish this with the node maintenance window suggestions (which I do not think is possible), I don't think that should be the accepted solution here because the node maintenance window settings seem to imply that there will be additional functionality changes above and beyond the PDB removal which is not desired for the use cases I have outlined above.

**Comment by codenightlong** [comment]
We have a similar issue. We are using synchronous replication with local storage and primary going down is fine for us since some secondary will take over it. Due to PDB, our pod get stuck in "Pending" state and needs manual intervention to fix this. Since, some of our customers are dark site, its a problem for us to manually fix this. I would like the PDB to be toggled on or off.

**Comment by jsilvela** [comment]
Closing with the merge of https://github.com/cloudnative-pg/cloudnative-pg/pull/3837

**Comment by jsilvela** [comment]
https://github.com/cloudnative-pg/cloudnative-pg/pull/3837

**Comment by pgodowski** [comment]
Can we please ask to backport this into 1.22 which is the current LTS version?
I think it was supposed to be included in 1.22 as per @gbartolini 's comment [PR comment](https://github.com/cloudnative-pg/cloudnative-pg/pull/3837#issuecomment-1986048216), but I don't see it in 1.22.

**Comment by gbartolini** [comment]
> Can we please ask to backport this into 1.22 which is the current LTS version? I think it was supposed to be included in 1.22 as per @gbartolini 's comment [PR comment](https://github.com/cloudnative-pg/cloudnative-pg/pull/3837#issuecomment-1986048216), but I don't see it in 1.22.

Doing.
